src/backend/executor/execMain.c

```diff
@@ -4603,10 +4604,27 @@ targetid_get_partition(Oid targetid, EState *estate, bool openIndices)
 		if (openIndices)
 			ExecOpenIndices(childInfo);
 
-		map_part_attrs(parentInfo->ri_RelationDesc,
-					   childInfo->ri_RelationDesc,
-					   &(childInfo->ri_partInsertMap),
-					   TRUE); /* throw on error, so result not needed */
+		/*
+		 * If we are currently modifying a leaf partition, i.e we called
+		 * a DML command straight on child partition, or it's inheritance
+		 * plan execution (e.g. UPDATE command on root, which leads to multiple
+		 * subplans, and es_result_relations points to one of the partitions),
+		 * it will be better to omit building an ri_partInsertMap. In both
+		 * planners when we are executing a DML just on leaf partition,
+		 * the tuple descriptor already matches the partition's, and the extra
+		 * mapping in unnecessary (besides the map from makePartitionCheckMap,
+		 * which only helps to select partition in slot_get_partition).
+		 * Otherwise, there is a chance to reconstruct already valid tuple and
+		 * get the wrong results (e.g. target partition relation descriptor is
+		 * different from parentInfo's, but it's UPDATE (legacy planner) and
+		 * parentInfo represents another partition, which is not the true
+		 * parent).
+		 */
+		if (!currRelInfo || RelationGetRelid(currRelInfo->ri_RelationDesc) != targetid)
```

Comment:
in which case can `currRelInfo` be null and why should we do `map_part_attrs` in this case?

Reply:
Yep, there is no case for it to be NULL. However, after short exploration the usages of `targetid_get_partition` function (it is also used in case of COPY cmd), I can conclude that the issue related to these changes is only for UPDATE command for planner case (for inheritance case). Therefore, I'll reconsider that part of code and probably move it to another place.

Reply:
If we want to prevent the `ri_partInsertMap` construction, we can only make changes inside the `targetid_get_partition` function. Otherwise, for example, if we make such validation after the partition is selected (at execInsert), there would be no option except the `pfree(resultRelInfo->ri_partInsertMap), resultRelInfo->ri_partInsertMap = NULL ` call.

Therefore, I only changed the map building condition, which allows to create the map only when the DML is executed on root partition.